core/vm: less allocations for various call variants

holiman opened the pull request:
This PR is a bit experimental, and contains a lot of reductions in allocations. It's based off of five different benchmarks

- Staticcall to identity precompile.
- Call to identity precompile, 
- A pretty 'normal' loop which does nothing other than loop (no calls)
- A 0-value call to an EOA (no code)
- A 0-value call to a contract which immediately reverts

In most cases, allocations could be brought down to a constant, exceptions being call to nonexist and call to reverting, where allocs scale linearly with the number of calls made (however, this PR still reduces the amount by a lot). 

```
[user@work go-ethereum]$ benchstat before.txt after.txt 
name                                   old time/op    new time/op    delta
SimpleLoop/staticcall-identity-100M-6     115ms ± 1%      62ms ± 1%   -46.25%  (p=0.008 n=5+5)
SimpleLoop/call-identity-100M-6           281ms ± 2%      81ms ± 6%   -71.28%  (p=0.008 n=5+5)
SimpleLoop/loop-100M-6                    690ms ± 4%     679ms ± 2%      ~     (p=0.548 n=5+5)
SimpleLoop/call-nonexist-100M-6           459ms ±49%     162ms ± 4%   -64.60%  (p=0.008 n=5+5)
SimpleLoop/call-EOA-100M-6                570ms ± 9%      73ms ± 1%   -87.28%  (p=0.008 n=5+5)
SimpleLoop/call-reverting-100M-6          3.30s ±60%     0.17s ± 1%   -94.90%  (p=0.016 n=5+4)

name                                   old alloc/op   new alloc/op   delta
SimpleLoop/staticcall-identity-100M-6    42.9MB ± 0%     0.0MB ± 2%   -99.99%  (p=0.008 n=5+5)
SimpleLoop/call-identity-100M-6          42.7MB ± 0%     0.0MB ± 2%   -99.99%  (p=0.008 n=5+5)
SimpleLoop/loop-100M-6                   14.0kB ± 0%     4.2kB ±18%   -70.28%  (p=0.008 n=5+5)
SimpleLoop/call-nonexist-100M-6           105MB ± 0%      15MB ± 0%   -85.41%  (p=0.008 n=5+5)
SimpleLoop/call-EOA-100M-6               43.6MB ± 0%     0.0MB ± 0%   -99.99%  (p=0.016 n=5+4)
SimpleLoop/call-reverting-100M-6         1.24GB ± 0%    0.06GB ± 0%   -95.54%  (p=0.008 n=5+5)

name                                   old allocs/op  new allocs/op  delta
SimpleLoop/staticcall-identity-100M-6      670k ± 0%        0k ± 0%   -99.99%  (p=0.008 n=5+5)
SimpleLoop/call-identity-100M-6            668k ± 0%        0k ± 0%   -99.99%  (p=0.008 n=5+5)
SimpleLoop/loop-100M-6                     42.0 ± 0%      31.0 ± 6%   -26.19%  (p=0.008 n=5+5)
SimpleLoop/call-nonexist-100M-6           1.09M ± 0%     0.27M ± 0%   -75.00%  (p=0.008 n=5+5)
SimpleLoop/call-EOA-100M-6                 681k ± 0%        0k ± 0%  -100.00%  (p=0.008 n=5+5)
SimpleLoop/call-reverting-100M-6          1.62M ± 0%     1.08M ± 0%   -33.34%  (p=0.008 n=5+5)
```

holiman commented:
Benchmark: https://geth-bench.ethdevops.io/d/Jpk-Be5Wk/dual-geth?orgId=1&var-exp=bench03&var-master=bench04&var-percentile=50&from=1592141282530&to=now

holiman commented:
I took a memory dump from this PR as it was executing. 

![profile001](https://user-images.githubusercontent.com/142290/84655431-3a8f5c00-af11-11ea-8bcc-6fa98c0cb31d.png)

holiman commented:
On the benchmark run, this PR is ahead by `35m` after ~24 hours (https://geth-bench.ethdevops.io/d/Jpk-Be5Wk/dual-geth?orgId=1&var-exp=bench03&var-master=bench04&var-percentile=50&from=1592227036577&to=1592229862284)

karalabe commented on core/state/statedb.go:
```diff
@@ -540,13 +544,14 @@ func (s *StateDB) getDeletedStateObject(addr common.Address) *stateObject {
 			return nil
 		}
-		if err := rlp.DecodeBytes(enc, &data); err != nil {
+		data = &Account{}
```
Pls. use `data = new(Account)`

karalabe commented on core/state/statedb.go:
```diff
@@ -540,13 +544,14 @@ func (s *StateDB) getDeletedStateObject(addr common.Address) *stateObject {
 	}
 	// Insert into the live set
-	obj := newObject(s, addr, data)
+	obj := newObject(s, addr, *data)
```
It feels a bit weird to dereference `data` here. Can't we just pass the pointer and make stateObject store a pointer too?

karalabe replied:
CC @fjl

holiman replied:
@fjl do you have any comments about this? 
I preferred initially to keep the changes as small as possible, thus didn't change the referencing

holiman commented on core/state/statedb.go:
```diff
@@ -718,6 +723,17 @@ func (s *StateDB) RevertToSnapshot(revid int) {
 	s.validRevisions = s.validRevisions[:idx]
 }
 
+func (s *StateDB) DiscardSnapshot(revid int) {
+	// Find the snapshot in the stack of valid snapshots.
+	idx := sort.Search(len(s.validRevisions), func(i int) bool {
+		return s.validRevisions[i].id >= revid
```
Pop off instead, but check that the revision id is correct. If not ,leave it there

karalabe commented on core/vm/evm.go:
```diff
@@ -41,23 +42,23 @@ type (
+		precompiles = PrecompiledContractsByzantium
+	} else {
+		precompiles = PrecompiledContractsHomestead
+	}
```
Nitpick, pls use a switch :)

karalabe commented on core/vm/evm.go:
```diff
@@ -224,35 +217,50 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
 		defer func() { // Lazy evaluation of the parameters
-			evm.vmConfig.Tracer.CaptureEnd(ret, gas-contract.Gas, time.Since(start), err)
+			evm.vmConfig.Tracer.CaptureEnd(ret, startGas-gas, time.Since(start), err)
 		}()
```
defer func(startGas uint64) {...}(gas)

karalabe replied:
same with start

karalabe commented on core/vm/evm.go:
```diff
@@ -224,35 +217,50 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
+			ret, err = nil, nil // gas is unchanged
+		} else {
+			var addrCopy common.Address
+			addrCopy.SetBytes(addr.Bytes())
```
addrCopy := addr

holiman commented on core/vm/stack.go:
```diff
@@ -30,7 +37,13 @@ type Stack struct {
 
 func newstack() *Stack {
-	return &Stack{data: make([]uint256.Int, 0, 16)}
+	//return &Stack{data: make([]uint256.Int, 0, 16)}
```
delete

karalabe commented on core/vm/stack.go:
```diff
@@ -89,18 +102,19 @@ func (st *Stack) Print() {
+	data []uint32
 }
 
 func newReturnStack() *ReturnStack {
```
might as well make a sync.Pool for this too

holiman commented:
The bench run got stopped on a bad block -- I had kind of been expecting that: after the run was started, I added the commit `core/vm: tracer fix + staticcall gas fix `. Was curious to see how far it would get without that fix, stopped at block `9778717`. I updated to the latest docker image and started again, it continued without complains.

holiman commented:
This is roughly one day after they've both been synced with chain tip.

![Screenshot_2020-06-24 Dual Geth - Grafana](https://user-images.githubusercontent.com/142290/85577171-4ed20800-b639-11ea-8642-01c16adc2a06.png)

And allocs (yellow is this PR) : 
![Screenshot_2020-06-24 Dual Geth - Grafana(1)](https://user-images.githubusercontent.com/142290/85577278-63ae9b80-b639-11ea-9c85-5247f3eab046.png)

MariusVanDerWijden commented on core/vm/evm.go:
```diff
@@ -41,23 +42,24 @@ type (
+		precompiles = PrecompiledContractsHomestead
+	}
+	p, ok := precompiles[addr]
+	return p, ok
```
nitpick: return results directly

holiman replied:
Actually, that doesn't work: 
```diff
diff --git a/core/vm/evm.go b/core/vm/evm.go
index 4dd54ba5fb..0d8b283bfd 100644
--- a/core/vm/evm.go
+++ b/core/vm/evm.go
@@ -54,8 +54,9 @@ func (evm *EVM) precompile(addr common.Address) (PrecompiledContract, bool) {
        default:
                precompiles = PrecompiledContractsHomestead
        }
-       p, ok := precompiles[addr]
-       return p, ok
+       return precompiles[addr]
+       //p, ok := precompiles[addr]
+       //return p, ok
 }

 // run runs the given contract and takes care of running precompiles with a fallback to the byte code interpreter.
``` 
```
[user@work go-ethereum]$ make
env GO111MODULE=on go run build/ci.go install ./cmd/geth
>>> /usr/local/go/bin/go install -ldflags -X main.gitCommit=5bfbb7d986a4aa0e8f37931b6db76e67f5eef974 -X main.gitDate=20200715 -v ./cmd/geth
github.com/ethereum/go-ethereum/core/vm
# github.com/ethereum/go-ethereum/core/vm
core/vm/evm.go:57:2: not enough arguments to return
        have (PrecompiledContract)
        want (PrecompiledContract, bool)
util.go:45: exit status 2
```

MariusVanDerWijden commented on core/vm/evm.go:
```diff
@@ -277,23 +283,27 @@ func (evm *EVM) CallCode(caller ContractRef, addr common.Address, input []byte,
-	contract.SetCallCode(&addr, evm.StateDB.GetCodeHash(addr), evm.StateDB.GetCode(addr))
-
-	ret, err = run(evm, contract, input, false)
+	var snapshot = evm.StateDB.Snapshot()
```
If we execute a precompile we don't need to take a snapshot (and later revert),
because the precompiles are stateless right?

MariusVanDerWijden replied:
Maybe we can change logic s.th. we only take a snapshot (and revert) with a normal delegatecall

MariusVanDerWijden replied:
Ahh nvm saw your comment down below

MariusVanDerWijden replied:
Well maybe not... the comment down below only talks about staticcall not precompiles

holiman replied:
We might get away with not taking a snapshot for a callcode to a precompile (also delegatecall, staticcall). However, there's no future-guarantee that all precompiles will never modify state. I didn't add that particular optimization, I'd rather save that for a future PR if it makes any difference (I think it doesn't really matter)

MariusVanDerWijden commented on core/vm/instructions.go:
```diff
@@ -563,7 +563,7 @@ func opJumpSub(pc *uint64, interpreter *EVMInterpreter, callContext *callCtx) ([
 		return nil, ErrInvalidJump
 	}
-	callContext.rstack.push(*pc)
+	callContext.rstack.push(uint32(*pc))
```
In the yellowpaper the pc is defined as Uint256, we could probably use uint32 though

karalabe commented on core/vm/contracts.go:
```diff
@@ -102,12 +102,18 @@ var PrecompiledContractsYoloV1 = map[common.Address]PrecompiledContract{
-	return nil, ErrOutOfGas
+	suppliedGas -= gasCost
+	output, err := p.Run(input)
+	return output, suppliedGas, err
```
Nit, imho the code would be clearer if we didn't do `suppliedGas -= gasCost` and then `return suppliedGas`, rather did `return suppliedGas - gasCost`

karalabe commented on core/vm/evm.go:
```diff
@@ -224,35 +218,47 @@ func (evm *EVM) Call(caller ContractRef, addr common.Address, input []byte, gas
+			contract := NewContract(caller, AccountRef(addrCopy), value, gas)
+			contract.SetCallCode(&addrCopy, evm.StateDB.GetCodeHash(addrCopy), code)
+			ret, err = run(evm, contract, input, false)
+			// When an error was returned by the EVM or when setting the creation code
```
I think this comment refers to the error handling 2 nesting outside at line 252. Shouldn't it be moved out?
